### DISCUSSION TITLE
Server properties

### DISCUSSION THREAD

**Discussion by alxtr1bal** [scottoffen/grapevine#50] [Q&A] [answered]
Hi, in GV5, is it still possible to do : 

```csharp
var prop = context.Server.GetPropertyValueAs<XXXX>("Prop");
```

I couldn't find anything like that...

**Comment by scottoffen** [comment, marked as answer]
These are called [`Locals`](https://github.com/scottoffen/grapevine/blob/main/src/Grapeseed/Locals.cs) now instead of `DynamicProperties`, to match the lingo used in ExpressJS from whence they are insipred. Locals are available on the `IRestServer` interface and the `IHttpContext` interface.

```csharp
string prop1 = server.Locals.GetAs("Prop1");
string prop2 = context.Locals.GetAs("Prop2");
// or
var prop3 = server.Locals.GetAs<string>("Prop3");
var prop4 = context.Locals.GetAs<string>("Prop4");
```

However, as of the latest release, the `IHttpContext` interface does not contain an `IRestServer` property. Which is to say that the `context.Server` property you reference above doesn't exist anymore. But you can inject it into your class, and you already get this for free!

```csharp
[RestResource]
public class MyResource
{
    private IRestServer _server;

    public MyResource(IRestServer server)
    {
        _server = server;
    }

    [RestRoute("Get", "/prop")]
    public async Task GetCookie(IHttpContext context)
    {
        var serverProp = _server.Locals.GetAs<string>("ServerPropertyName");
        var ctxProp = context.Locals.GetAs<string>("ContextPropertyName");
    }
}
```

Because the instance of `IRestServer` is already registered in the service collection used by the router (this happens automatically) as a singleton, you don't need to do any more wiring up for this to work.

The intention is that this approach will allow you to attach any kind of arbitrary data to anything that has a Locals property on it, and retrieve it later. In the case of the server, you can attach anything that is specific to the server. In the case of the `IHttpContext`, this is so that you can pass already parsed or calculated data between routes seamlessly.

**Reply by alxtr1bal** [reply]
Thank you very much for this :-)

**Reply by scottoffen** [reply]
This is going to change again in the next release. I'll update my examples after that is published.

**Reply by cbritton** [reply]
I want to pass in a custom class to a RestResource when the class is constructed.  Do you have examples of using Locals and attaching it to a server?

**Reply by scottoffen** [reply]
For custom classes passed into the RestResource, just use dependency injection.

```csharp
public class Startup
{
    public void ConfigureServices(IServiceCollection services)
    {
        // Inject your dependency here. If that class has any dependencies, add them, too!
        services.AddScoped<IMyClass, MyClass>();
    }
}
```

Make sure to wire up all your dependencies. They can be transient or singleton, too. Then you can easily inject them into your resource as needed.

```csharp
[RestResource]
public class MyResource
{
    private IMyClass _myClass;

    public MyResource(IMyClass myClass)
    {
        _myClass = myClass;
    }

    // Any routes you add will have access to the instance of your custom object.
}
```
